pkg/controller/nodeipam/ipam/multicidrset/cidr_set.go

```diff
@@ -0,0 +1,363 @@
+	}
+	s.Lock()
+	defer s.Unlock()
+
+	for i := begin; i <= end; i++ {
+		// Remove from the allocated CIDR Map and decrement the counter only if currently
+		// marked allocated. Avoids double counting.
+		currCIDR := s.indexToCIDRBlock(i)
+		if _, ok := s.AllocatedCIDRMap[currCIDR.String()]; ok {
+			delete(s.AllocatedCIDRMap, currCIDR.String())
+			s.allocatedCIDRs--
+			cidrSetReleases.WithLabelValues(s.label).Inc()
+		}
+	}
+
+	cidrSetUsage.WithLabelValues(s.label).Set(float64(s.allocatedCIDRs) / float64(s.maxCIDRs))
+
+	return nil
+}
+
+// Occupy marks the given CIDR range as used. Occupy succeeds even if the CIDR
+// range was previously used.
+func (s *CidrSet) Occupy(cidr *net.IPNet) (err error) {
+	begin, end, err := s.getBeginningAndEndIndices(cidr)
+	if err != nil {
+		return err
+	}
+	s.Lock()
+	defer s.Unlock()
+
+	for i := begin; i <= end; i++ {
+		// Add to the allocated CIDR Map and increment the counter only if not already
+		// marked allocated. Avoids double counting.
+		currCIDR := s.indexToCIDRBlock(i)
+		if _, ok := s.AllocatedCIDRMap[currCIDR.String()]; !ok {
+			s.AllocatedCIDRMap[currCIDR.String()] = true
+			cidrSetAllocations.WithLabelValues(s.label).Inc()
+			s.allocatedCIDRs++
+		}
+	}
+	cidrSetUsage.WithLabelValues(s.label).Set(float64(s.allocatedCIDRs) / float64(s.maxCIDRs))
+
+	return nil
+}
+
+func (s *CidrSet) getIndexForCIDR(cidr *net.IPNet) (int, error) {
+	return s.getIndexForIP(cidr.IP)
+}
+
+func (s *CidrSet) getIndexForIP(ip net.IP) (int, error) {
+	if ip.To4() != nil {
+		cidrIndex := (binary.BigEndian.Uint32(s.clusterCIDR.IP) ^ binary.BigEndian.Uint32(ip.To4())) >> uint32(32-s.nodeMaskSize)
+		if cidrIndex >= uint32(s.maxCIDRs) {
+			return 0, fmt.Errorf("CIDR: %v/%v is out of the range of CIDR allocator", ip, s.nodeMaskSize)
+		}
+		return int(cidrIndex), nil
+	}
+	if ip.To16() != nil {
+		bigIP := big.NewInt(0).SetBytes(s.clusterCIDR.IP)
+		bigIP = bigIP.Xor(bigIP, big.NewInt(0).SetBytes(ip))
+		cidrIndexBig := bigIP.Rsh(bigIP, uint(net.IPv6len*8-s.nodeMaskSize))
+		cidrIndex := cidrIndexBig.Uint64()
+		if cidrIndex >= uint64(s.maxCIDRs) {
+			return 0, fmt.Errorf("CIDR: %v/%v is out of the range of CIDR allocator", ip, s.nodeMaskSize)
+		}
+		return int(cidrIndex), nil
+	}
+
+	return 0, fmt.Errorf("invalid IP: %v", ip)
+}
+
+// Contains checks whether a CIDR contains an IP address.
+func (s *CidrSet) Contains(ip net.IP) bool {
+	return s.clusterCIDR.Contains(ip)
+}
+
+// CheckIntersect checks whether a CIDR intersects with another CIDR.
+func (s *CidrSet) CheckIntersect(c *CidrSet) bool {
+	return s.clusterCIDR.Contains(c.clusterCIDR.IP)
+}
+
+// CIDRLabel is a getter function for cidrSet label.
+func (s *CidrSet) CIDRLabel() string {
+	return s.label
+}
+
+// HasAllocated checks whether a CIDRset has CIDRs allocated.
+func (s *CidrSet) HasAllocated() bool {
+	return s.allocatedCIDRs != 0
+}
+
+// MaxCIDRs returns maximum number of CIDRs that can be allocated.
+func (s *CidrSet) MaxCIDRs() int {
```

Comment:
any reason why we use accessors instead of exposing fields?

Reply:
Accessors are not required here, the fields can be exposed. Wasn't sure initially if these fields should be exposed